### PR TITLE
types: fix delete error when convert string to float or int

### PR DESCRIPTION
**Pull request by DQinYuan**
### What problem does this PR solve? 

To fix the issue    [#10806](https://github.com/pingcap/tidb/issues/10806)

When delete from a varchar column only with empty string and a condition compare with a integer or float, tidb will return a "Data Truncated Error" and mysql(both 5.x and 8.x) will only return "Query OK, 0 rows affected (0.01 sec) "

### What is changed and how it works?

Before:

```sql
tidb>create table t (a varchar(30));
tidb>insert into t values ('');
tidb>delete from t where a=10;
ERROR 1265 (01000): Data Truncated
```

After:

```sql
tidb>create table t (a varchar(30));
tidb>insert into t values ('');
tidb>delete from t where a=10;
Query OK, 0 rows affected (0.01 sec) 
```

modify method `getValidFloatPrefix` in types/convert.go

When stmt context is "delete"(i.e. sc.InDeleteStmt is true) and string is empty, `getValidFloatPrefix` directly return "0" and nil error.

```go
func getValidFloatPrefix(sc *stmtctx.StatementContext, s string) (valid string, err error) {
	if sc.InDeleteStmt && s == "" {
		return "0", nil
	}
......
```

### Check List 
Tests 

 - Unit test

### CODE REVIEW

File: types/convert.go [hunk 1]
```diff
@@ -580,6 +580,10 @@ func ConvertJSONToDecimal(sc *stmtctx.StatementContext, j json.BinaryJSON) (*MyD
 
 // getValidFloatPrefix gets prefix of string which can be successfully parsed as float.
 func getValidFloatPrefix(sc *stmtctx.StatementContext, s string) (valid string, err error) {
+	if sc.InDeleteStmt && s == "" {
```
Comment (winkyao) [inline types/convert.go]: This function is used in `StrToUint`, maybe not only delete statement?

Comment (DQinYuan) [reply types/convert.go]: The  condition  `statementContext.InDeleteStmt == true` will keep it only work in delete statement. 

And in delete statement, empty string can always convert to 0 without error and warning.(according to behaviour of mysql 5.x and 8.x ).

But in other statement like select, tidb is consistent with mysql(all with warning),  so I restrict it in delete statement context.

**Comment by DQinYuan** [comment]
@Deardrops   You are right, I have corrected unit test in this pr.

**Comment by Deardrops** [comment]
LGTM

**Comment by zz-jason** [comment]
@DQinYuan Sorry for the late review. Could you update the PR with the latest master branch and resolve the conflicts?

**Pull request merged**